Re-implement auth login flow (#1368) with tiny-http

lsegal opened the pull request:
Experiment using tiny-http instead of actix-web to implement #1368

brynary reviewed (commented):
Nice! Thanks for implementing this. I added some comments, mostly around error handling.

Overall, it feels like a bit more code, offset by being lighter from a dependency standpoint.

My gut feel is that while we just have one endpoint this is a good place to be. If/when we extend the CLI to providing interactive web UI functionality, we probably should re-evaluate with a bias towards something more full featured like actix

lsegal commented on qlty-cloud/src/auth/auth_flow.rs:
```diff
@@ -99,53 +76,91 @@ fn redirect_matches(expected: String, actual: String) -> bool {
+    }
+
+    let (shutdown_send, shutdown_recv): (Sender<()>, Receiver<()>) = mpsc::channel();
+    let server = server.unwrap();
```
There's no panic risk here because of the err check above. That might be non-obvious though.

lsegal commented on qlty-cloud/src/auth/auth_flow.rs:
```diff
@@ -99,53 +76,91 @@ fn redirect_matches(expected: String, actual: String) -> bool {
+pub fn launch_login_server(state: AppState) -> Result<ServerResponse> {
+    let server = Server::http("127.0.0.1:0");
+    if let Err(err) = server {
+        bail!("Failed to create server: {}", err);
```
```
the method `with_context` exists for enum `Result<Server, Box<dyn Error + Send + Sync>>`, but its trait bounds were not satisfied
the following trait bounds were not satisfied:
`Box<dyn StdError + Send + std::marker::Sync>: anyhow::context::ext::StdError`
which is required by `Result<tiny_http::Server, Box<dyn StdError + Send + std::marker::Sync>>: anyhow::Context<tiny_http::Server, Box<dyn StdError + Send + std::marker::Sync>>`
```

This is typically the error you get when you try to manipulate this error object, since it's not a regular error and doesn't implement Sized. I think we can simplify by using `map_error` but that would still lose "context".

lsegal commented on qlty-cloud/src/auth/auth_flow.rs:
```diff
@@ -99,53 +76,91 @@ fn redirect_matches(expected: String, actual: String) -> bool {
-            .unwrap()
-            .block_on(server);
-    });
+    let token = Token::new(&state.credential_user);
```
Calling it TokenStore would defeat the purpose of abstracting storage. At that point we could just dump the struct entirely and use regular methods to read/write from storage-- I had an initial version with this but removed it to avoid changing too much of the original implementation.

Ideally we would just have a simplified function oriented API:

```rust
// mod credential_store
fn read_token(user: &str) -> Result<String>
fn write_token(user: &str, value: &str) -> Result<()>
```

There's really no reason to create a token object / hydrate values since it's not used in any long-lived fashion. We throw away the Token object as soon as we read: https://github.com/qltysh/qlty/pull/1368/files#diff-62de191f887ff041d4cf140614b940a83d994dcde45fe5763f2063ebd076cf44R14-R18

brynary replied:
Ah yes, that makes sense. Agreed.

lsegal commented on qlty-cloud/src/auth/auth_flow.rs:
```diff
@@ -157,77 +172,92 @@ mod tests {
 
-        let token = Token::new("qlty-cli-test").get().unwrap();
-        assert_eq!(token, "ABCDEFG");
+        if !cfg!(target_os = "macos") {
```
macOS prompts for a password to read/write to token storage, presumably this will break tests on CI

lsegal commented:
@brynary addressed comments. I'm fine with the approach of using this for now, but note that there are a lot of dragons and hidden complexity with this implementation and will likely hinder us from making larger refactors, even if it is just one API endpoint.
